proto/go_proto_library.bzl

```diff
@@ -53,95 +53,83 @@ def _go_prefix(ctx):
     prefix = prefix + "/"
   return prefix
 
-def _external_dir(file):
-  """Returns the include dir for external files or empty.
-
-  file: a bazel File object
-
-  Returns:
-    a string path or "" if nothing to do
-
-  Examples to look for
-    'external/some_repo'
-    'bazel-$project/external/some_repo'
-    'bazel-out/local-fastbuild/genfiles/external/some_repo'
-    'bazel-out/local-fastbuild/genfiles/'
-"""
-  toks = file.dirname.split("/")
-  if toks[0] == "external":
-    return "/".join(toks[:2])
-  if not toks[0].startswith("bazel-"):
-    return ""
-  for idx in range(1, len(toks)):
-    if toks[idx] == "external":
-      return "/".join(toks[:idx+2])
-  return "/".join(toks[:3])
-
-def _external_dirs(files):
-  """Compute any needed -I options to protoc from external filegroups."""
-  res = []
-  for f in files:
-    e = _external_dir(f)
-    if e:
-      res.append(e)    
-  return set(res)
-
-def _go_proto_library_gen_impl(ctx):
-  """Rule implementation that generates Go using protoc."""
-  bazel_style = ctx.label.name != _DEFAULT_LIB + _PROTOS_SUFFIX
-  protos = list(ctx.files.srcs)
-  go_package_name = ""
-  if bazel_style:
-    go_package_name = "/" + ctx.label.name[:-len(_PROTOS_SUFFIX)]
-  m_import_path = ",".join(["M%s=%s%s%s" % (f.path, _go_prefix(ctx),
-                                            ctx.label.package, go_package_name)
-                            for f in ctx.files.srcs])
+def _collect_protos_import(ctx):
+  """Collect the list of transitive protos and m_import_path."""
+  protos = set()
+  m_import_path = ""
   for d in ctx.attr.deps:
     if not hasattr(d, "_protos"):
       # should be a raw filegroup then
       protos += list(d.files)
       continue
     protos += d._protos
     m_import_path += "," + d._m_import_path
-  use_grpc = ""
-  if ctx.attr.grpc:
-    use_grpc = "plugins=grpc,"
+  return list(protos), m_import_path
+
+def _drop_external(path):
+  """Drop leading '../' indicating an external dir of the form ../$some-repo."""
+  if path[:3] != "../":
```

Comment:
I would prefer to use startswith (https://www.bazel.build/versions/master/docs/skylark/lib/string.html#startswith) to make this more easily readable.